src/python/pants/build_graph/build_configuration.py

```diff
@@ -5,203 +5,245 @@
 import typing
 from collections.abc import Iterable
 from dataclasses import dataclass, field
-from typing import Any, Dict, Type, Union
+from typing import Any, Callable, Dict, Type, Union
 
 from pants.build_graph.build_file_aliases import BuildFileAliases
-from pants.engine.rules import RuleIndex
+from pants.engine.rules import Rule, RuleIndex
 from pants.engine.target import Target
 from pants.option.optionable import Optionable
-from pants.util.ordered_set import OrderedSet
+from pants.util.frozendict import FrozenDict
+from pants.util.ordered_set import FrozenOrderedSet, OrderedSet
 
 logger = logging.getLogger(__name__)
 
 
-@dataclass
+@dataclass(frozen=True)
 class BuildConfiguration:
```

Comment:
Aplogies for this diff - this GitHub diff viewer does not like it at all. This was just the following refactor:
1. Add @dataclass(frozen=True) to BuildConfiguration and lift up _variables to hold the immutable data and moving all the accessors of that data to the top of the file.
2. Indent all the mutators and enclose them in the inner dataclass Builder.
3. Add a create method.